### ISSUE TITLE
memory leak of tf2.3 with LSTM

### ISSUE THREAD

**Issue by hlya23dd** [tensorflow/tensorflow#46633]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Red Hat Enterprise Linux Server 7.9 (Maipo)
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7.4

**Describe the current behavior**
memory leak with training batch when using LSTM layer (no problem with dense `layer)`
**Describe the expected behavior**
no memory leak
**Standalone code to reproduce the issue**
      
import numpy as np
import tensorflow
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.callbacks import Callback

import resource
class MemoryCallback(Callback):
    def on_epoch_end(self, epoch, log={}):
        print('')
        print('Memory usage: ', resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
        
if __name__ == "__main__":
        n_sample=1000
        n_timestep =1000
        num_channel = 1
        batch_size=100
        epochs=10
        i_train_on_batch = False
        
        train = np.random.rand(n_sample, n_timestep, num_channel)
        label = np.random.rand(n_sample, num_channel)
        
        x=layers.Input(shape=(n_timestep,num_channel))
        z=  layers.LSTM(num_channel)(x)  
        model = keras.Model(inputs=x, outputs=z)
        
        model.compile(optimizer="adam", loss="mse")
        if i_train_on_batch:
           for epoch in range(epochs):
               for step in range(n_sample//batch_size):
                   model.train_on_batch(train[step*batch_size:(step+1)*batch_size], label[step*batch_size:(step+1)*batch_size],)
               print('Epoch ', epoch,'/', epochs, '  Memory usage: ', resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
        else:
           model.fit(train, label, batch_size=batch_size, epochs=epochs, callbacks=[MemoryCallback()])

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

###################################################
i_train_on_batch = False:
###################################################

bash-4.2$ python memory_leak_epoch.py 
2021-01-23 16:17:38.709232: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-23 16:17:43.397736: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2021-01-23 16:17:43.404166: E tensorflow/stream_executor/cuda/cuda_driver.cc:314] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2021-01-23 16:17:43.404195: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (taurusi4181.taurus.hrsk.tu-dresden.de): /proc/driver/nvidia/version does not exist
2021-01-23 16:17:43.406436: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
Epoch 1/10
10/10 [==============================] - ETA: 0s - loss: 0.2690
Memory usage:  424672
10/10 [==============================] - 4s 364ms/step - loss: 0.2690
Epoch 2/10
10/10 [==============================] - ETA: 0s - loss: 0.2612
Memory usage:  475888
10/10 [==============================] - 4s 370ms/step - loss: 0.2612
Epoch 3/10
10/10 [==============================] - ETA: 0s - loss: 0.2533
Memory usage:  527780
10/10 [==============================] - 4s 374ms/step - loss: 0.2533
Epoch 4/10
10/10 [==============================] - ETA: 0s - loss: 0.2453
Memory usage:  578732
10/10 [==============================] - 4s 376ms/step - loss: 0.2453
Epoch 5/10
10/10 [==============================] - ETA: 0s - loss: 0.2373
Memory usage:  627480
10/10 [==============================] - 4s 381ms/step - loss: 0.2373
Epoch 6/10
10/10 [==============================] - ETA: 0s - loss: 0.2293
Memory usage:  679224
10/10 [==============================] - 4s 386ms/step - loss: 0.2293
Epoch 7/10
10/10 [==============================] - ETA: 0s - loss: 0.2212
Memory usage:  728328
10/10 [==============================] - 4s 385ms/step - loss: 0.2212
Epoch 8/10
10/10 [==============================] - ETA: 0s - loss: 0.2130
Memory usage:  777960
10/10 [==============================] - 4s 385ms/step - loss: 0.2130
Epoch 9/10
10/10 [==============================] - ETA: 0s - loss: 0.2050
Memory usage:  827592
10/10 [==============================] - 4s 390ms/step - loss: 0.2050
Epoch 10/10
10/10 [==============================] - ETA: 0s - loss: 0.1969
Memory usage:  877752
10/10 [==============================] - 4s 388ms/step - loss: 0.1969
bash-4.2$ 
###################################################
i_train_on_batch = True:
###################################################

bash-4.2$ python memory_leak_epoch.py 
2021-01-23 16:15:53.238289: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2021-01-23 16:15:58.148876: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2021-01-23 16:15:58.155607: E tensorflow/stream_executor/cuda/cuda_driver.cc:314] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2021-01-23 16:15:58.155647: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (taurusi4181.taurus.hrsk.tu-dresden.de): /proc/driver/nvidia/version does not exist
2021-01-23 16:15:58.158445: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
Epoch  0 / 10   Memory usage:  419684
Epoch  1 / 10   Memory usage:  472220
Epoch  2 / 10   Memory usage:  523804
Epoch  3 / 10   Memory usage:  574228
Epoch  4 / 10   Memory usage:  623068
Epoch  5 / 10   Memory usage:  673756
Epoch  6 / 10   Memory usage:  724180
Epoch  7 / 10   Memory usage:  774076
Epoch  8 / 10   Memory usage:  824500
Epoch  9 / 10   Memory usage:  874660

**Comment by amahendrakar** [comment]
@hlya23dd,
I was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/2a8bbf5ed11817e2eae34066f37472e0/46633-2-3.ipynb) and [TF v2.4](https://colab.research.google.com/gist/amahendrakar/da77117dbfde3cbc3328a2da29a9ed5c/46633.ipynb).

However on running the code with the latest [TF-nightly](https://colab.research.google.com/gist/amahendrakar/dbf2b2346cd5c41f3afa1c8ccde18f3c/46633-tf-nightly.ipynb), the memory increases for the first few epochs and then remains stable. Please check the linked gist for reference. Thanks!
